### PR TITLE
supplemental #14872: test cases

### LINKED ISSUE #14872: UnnecessaryParentheses: no violation if casting is present before expression
detected at https://github.com/checkstyle/checkstyle/pull/14856/files#r1597531069

I have read check documentation: https://checkstyle.org/checks/coding/unnecessaryparentheses.html#UnnecessaryParentheses
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```bash
$ cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
    <module name="UnnecessaryParentheses">
    </module>
  </module>
</module>

$ cat Test.java 
import java.util.Arrays;

public class Test
{
  public static void main(String[] args) {
    String[] a = { "s", "", "1", "", "" };
    Arrays.stream(a).filter(s -> !((boolean)s.isEmpty())).toArray(String[]::new); // should be violation
    Arrays.stream(a).filter(s -> !(boolean)s.isEmpty()).toArray(String[]::new);
  }
}

$ java -jar checkstyle-10.16.0-all.jar -c config.xml Test.java
Starting audit...
Audit done.
```

Expected: violation on  line 7

--------

https://introcs.cs.princeton.edu/java/11precedence/

>Most programmers do not memorize them all, and, even those that do, use parentheses for clarity.

higher precedence executes first.

|Level|Operator                              |Description                                                                                       |Associativity|
|-----|--------------------------------------|--------------------------------------------------------------------------------------------------|-------------|
|16   |() [] new . ::                        |parentheses, array access,  object creation, member access, method reference                           |left-to-right|
|15   |=++ --                                |unary post-increment, unary post-decrement                                                         |left-to-right|
|14   |=+ - ! ~ ++ --                        |unary plus, unary minus, unary logical NOT, unary bitwise NOT, unary pre-increment, unary pre-decrement|right-to-left|
|13   |()                                    |cast                                                                                              |right-to-left|
|12   |* / %                                 |multiplicative                                                                                    |left-to-right|
|11   |=+ - +                                |additive, string concatenation                                                                     |left-to-right|
|10   |<< >> >>>                             |shift                                                                                             |left-to-right|
|9    |< <= > >= instanceof                  |relational                                                                                        |left-to-right|
|8    |== !=                                 |equality                                                                                          |left-to-right|
|7    |&                                     |bitwise AND                                                                                       |left-to-right|
|6    |^                                     |bitwise XOR                                                                                       |left-to-right|
|5    |&#124;                                     |bitwise OR                                                                                        |left-to-right|
|4    |&&                                    |logical AND                                                                                       |left-to-right|
|3    |&#124;&#124;                                    |logical OR                                                                                        |left-to-right|
|2    |?:                                    |ternary                                                                                           |right-to-left|
|1    |= += -= *= /= %= &= ^= &#124;= <<= >>= >>>=|assignment                                                                                        |right-to-left|
|0    |-> ->                                 |lambda expression, switch expression                                                               |right-to-left|

### PR DESCRIPTION
**Pull request by mohitsatr**
resolves #14872

### CODE REVIEW

**Comment by mohitsatr** [comment]
could you suggest better name for ```checkASTSurrounded```.

**Comment by mohitsatr** [comment]
refactoring is done to avoid NCSS and Cognitive Complexity warnings.

**Comment by mohitsatr** [comment]
Github, generate report for UnnecessaryParentheses/all-examples-in-one

**Comment by romani** [comment]
GitHub, generate report by config from src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesCasts.java

**Comment by romani** [comment]
Github, generate report for UnnecessaryParentheses/Example1

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesCasts.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+        int d = ((int) 100f) + 100 * 2 / ((int) 12.5f) + (int) 90f; // 2 violations
+
+        y = (int) (22.2 * 2) / ((int) 8f + 5);
+
+        double arg2 = 23.2;
+        int i = (int) arg2;
+
+        i = ((int) arg2); // violation 'Unnecessary parentheses around assignment right-hand side'
+
+        x = (2 * 2 /((int) k)); // 2 violations
```
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesCasts.java:32]: Please add violation message for both violations.
`// 2 violations below:`

Same for other cases like this

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesCasts.java [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+        float k = 12f;
+
+        int d = ((int) 100f) + 100 * 2 / ((int) 12.5f) + (int) 90f; // 2 violations
+
+        y = (int) (22.2 * 2) / ((int) 8f + 5);
+
+        double arg2 = 23.2;
+        int i = (int) arg2;
+
+        i = ((int) arg2); // violation 'Unnecessary parentheses around assignment right-hand side'
```
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesCasts.java:30]: Please put under each case with violation, same code but without violation.

**Review by romani — CHANGES_REQUESTED** [review]
Please add case https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/d9014bdb0943bd2a6f1e1d3a2e5f39a85659e771_2024000552/reports/diff/camunda/xref/home/runner/work/checkstyle/checkstyle/.ci-temp/repositories/camunda/operate/qa/data-generator/src/main/java/io/camunda/operate/data/generation/ResultChecker.java.html#L49
To inputs.

Major concern if we need to add this token to default or not.

Items:

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesCasts.java (hunk 1 above)
Comment (mohitsatr) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesCasts.java]: done

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesCasts.java (hunk 2 above)
Comment (mohitsatr) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesCasts.java]: done

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesCasts1.java [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+        float xy = 40f;
+        int yp = 0;
+        boolean finished = true;
+        boolean result = false;
+
+        if(py >= ((int)xy) // violation 'Unnecessary parentheses around expression.'
+                | (yp ==1 | py >=1)) {
+            xy--;
+        }
+
```
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesCasts1.java:53]: please remove empty line between same code with/without violation to  let them looks like block.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesCasts1.java [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+            checkone = !checkone;
+        }
+
+        double limit = 3.2;
+        for (int j = 0; j >= ((int) limit); j++) {
+            // violation above 'Unnecessary parentheses around expression.'
+            yp +=1;
+        }
+
+        for (int lp =0; lp >= (int) limit; lp++) {
```
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesCasts1.java:83]: remove empty line, and make it same (same variables) as previous FOR.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesCasts2.java [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    public void fooConditionals() {
+        int x = 23;
+        int y = 44;
+        float k = 12f;
+        boolean finished = true;
+        boolean result = false;
+
+        String filevalue = "FILEVALUE";
+        if (!finished
+                || !((boolean) filevalue.contains("O"))) {
```
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesCasts2.java:30]: make it single line as you did in next IF.

please apply same logic for other code in this file

**Review by romani — CHANGES_REQUESTED** [review]
>what impact would it have if we make it default ?

there will be build failures after users update to new version of checkstyle, if violation in unwanted users will be annoyed. 


items:

**Comment by romani** [comment]
please add this case to Inputs
`final ToIntFunction<T> comparison = ((Comparable<T>) element)::compareTo;` it is from https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/d9014bdb0943bd2a6f1e1d3a2e5f39a85659e771_2024000552/reports/diff/Vavr/index.html#A1

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesCasts2.java (hunk 5 above)
Comment (mohitsatr) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesCasts2.java]: done

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesCasts1.java (hunk 4 above)
Comment (mohitsatr) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesCasts1.java]: done

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesCasts1.java (hunk 3 above)
Comment (mohitsatr) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesCasts1.java]: done

**Comment by mohitsatr** [comment]
the following did show defect in implementations.
`final ToIntFunction<T> comparison = ((Comparable<T>) element)::compareTo;`
`assertEquals(1, ((int[]) bean.array)[0]);`
fixed now

**Comment by romani** [comment]
Github, generate report for UnnecessaryParentheses/Example1

**Comment by mohitsatr** [comment]
Presently, it logs unnecessary paren casting in ternary operators also. 
`f loat pin = ((Boolean) false) ? ((float) 21) : (float) 31; `

` throw (cause instanceof Exception) ? ((Exception) cause) : e; `

but it is unnecessary because exisiting code can easily detect the same thing with (non-default) `QUESTION` token. and it will log error even if user do not have QUESTION token enabled in the configs.
I'm fixing this.

**Comment by romani** [comment]
Please rebase over latest code to let CI pass.
Trigger diff report, to see changes.

**Comment by mohitsatr** [comment]
Github, generate report for UnnecessaryParentheses/Example1

**Comment by romani** [comment]
Github, generate report for UnnecessaryParentheses/Example1

**Comment by mohitsatr** [comment]
I have reviewed all the files (except the test files) in the report we generated a week ago. I have identified and addressed all the edge cases I could find, adding them along with other cases to the input. The remaining differences in this latest report are cases of unnecessary parentheses.

**Comment by mohitsatr** [comment]
`int j =  x == -1 ? -1 : ((int) 32.1) & 32;`

edge case that following logic could not detect 

https://github.com/checkstyle/checkstyle/blob/24eedce6a64145442db5b478c7a39ce0763dad8c/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnnecessaryParenthesesCheck.java#L465-L467

**Review by romani — COMMENTED** [review]
This report hard to verify without actual changes and compilation.

Do you have any ideas how we can be confident that all new violations in report are valid and extra characters can be simply removed.

Please add to Inputs case like 
https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/32de6f6a053018f117d4b9ce3febc4dd2b6b773e_2024171649/reports/diff/apache-struts/index.html#A1
With and without violation.

Cast with multiplication during placing to argument https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/32de6f6a053018f117d4b9ce3febc4dd2b6b773e_2024171649/reports/diff/apache-ant/index.html#A1
With and without violation.

>edge case that following logic could not detect

It is ok to not report violation.
We should avoid any false positives, complicated false negatives is ok to keep  with our violation

**Comment by mohitsatr** [comment]
> Do you have any ideas how we can be confident that all new violations in report are valid?

I have reviewed all the files (except the test files) in the latest report and have identified and addressed all the edge cases I could find, adding them along with other cases to the input. cases are similar and repeating. The remaining differences in report are related to cases of unnecessary parentheses.

**Comment by mohitsatr** [comment]
Github, generate report for UnnecessaryParentheses/Example2

**Comment by romani** [comment]
I become to worry that do validation without knowledge of precedence of evaluation/execution.
I updated with table:
https://github.com/checkstyle/checkstyle/issues/14872

precedence:
|Level|Operator                              |Description                                                                                       |Associativity|
|-----|--------------------------------------|--------------------------------------------------------------------------------------------------|-------------|
|16   |() [] new . ::                        |parentheses array access object creation member access method reference                           |left-to-right|
|15   |=++ --                                |unary post-increment unary post-decrement                                                         |left-to-right|
|14   |=+ - ! ~ ++ --                        |unary plus unary minus unary logical NOT unary bitwise NOT unary pre-increment unary pre-decrement|right-to-left|
|13   |()                                    |cast                                                                                              |right-to-left|


all operators above cast  in precedence can cause false positive, right ? 
It will compile with and without `(` and `)` but meaning will be different, so following checkstyle will cause defect.

**Comment by mohitsatr** [comment]
> all operators above cast in precedence can cause false positive, right ? 

yep, you are right. I completely overlooked this.

starting over....

**Comment by romani** [comment]
@mohitsatr , why you close PR?

**Comment by mohitsatr** [comment]
@romani I'm still figuring out this one. i will reopen as soon as find proper fix

**Comment by romani** [comment]
Let's merge at least tests, commit with prefix "supplemental: ....".
Implementation can be different, expected behavior should stay the same.
We did huge work to find out that this Checks might be broken, it has not precedence at all.

**Comment by romani** [comment]
@mohitsatr , if you think that logic is too complicated to be done, ok, let's abandon it for now, we can return to it later on.

Please keep only test updates, with comments "no violation until #14872" for lines where we would expect violation.
And keep comment "not extra parentheses due to operator precedence"

**Review by romani — APPROVED** [review]
thanks a lot !!!

at least test are contributed, it will help next steps for sure.
